src/reader.rs

```diff
@@ -102,8 +102,14 @@ pub struct Reader<R: BufRead> {
     /// Opened name start indexes into [`Self::opened_buffer`]. See documentation
     /// for that field for details
     opened_starts: Vec<usize>,
-    /// a buffer to manage namespaces
+
+    /// A buffer to manage namespaces
     ns_resolver: NamespaceResolver,
+    /// For `Empty` events keep the 'scope' of the namespace on the stack artificially. That way, the
+    /// consumer has a chance to use `resolve` in the context of the empty element. We perform the
+    /// pop as the first operation in the next `next()` call.
```

Comment:
The change makes sense generally, but just checking - the user could also avoid this whole scenario by enabling the expansion of Empty tags into Start and End tags, right? This is a workaround for this one specific scenario when that feature isn't enabled.

Reply:
I'm reading section 6.1 of https://www.w3.org/TR/xml-names11 which says:

>  The scope of a namespace declaration declaring a prefix extends from the beginning of the start-tag in which it appears to the end of the corresponding end-tag, excluding the scope of any inner declarations with the same NSAttName part. In the case of an empty tag, the scope is the tag itself. 

The way this reads to me is that providing this special case is a minor violation of the spec - by making the namespace of an empty tag available in this way, we are lying about the namespace scope.  

(Maybe you address that in later commits, I'm working my way through one by one)

Reply:
These lines was just moved from `NamespaceResolver`. This code does the right thing, because otherwise you will pop stored namespace before the user would able to use it:
```xml
<tag xmlns="..."/>
<!-- ^^^^^^^^^^^    - namespace introduced and pushed to the stack -->
<!--            ^^  - namespace popped from the stack -->
<!--              ^ - Event::Empty returned to the user, namespace is not availible -->
```
Compare to the events happening for the open-close pair:
```xml
<tag xmlns="...">
<!-- ^^^^^^^^^^^   - namespace introduced and pushed to the stack -->
<!--             ^ - Event::Start returned to the user, namespace is availible -->
</tag>
<!--^^             - namespace popped from the stack -->
<!--  ^            - Event::End returned to the user -->
```
Although `Event::End` is returned after its namespace was popped, `read_namespaced_event` returns that namespace as well, and you does not have attributes in the `End` tag where you could use the popped namespace.

When namespace popping is deferred, the situation is:
```xml
<tag xmlns="..."/>
<!-- ^^^^^^^^^^^    - namespace introduced and pushed to the stack -->
<!--              ^ - Event::Empty returned to the user, namespace is availible -->
<!--                - namespace popped from the stack on next event -->
```

That extending does not violates anything, because the user could access to that namespace only by calling methods on the returned `Event::Empty`, on the next call to `reader.next()` that namespace will be popped before returning the next event.

> the user could also avoid this whole scenario by enabling the expansion of Empty tags into Start and End tags, right?

Right.